image/rhel/konflux.Dockerfile

```diff
@@ -45,9 +45,8 @@ ENV MAIN_TAG_SUFFIX="$VERSIONS_SUFFIX" COLLECTOR_TAG_SUFFIX="$VERSIONS_SUFFIX" S
 
```

Comment:
IIUC:

Here are the places that BUILD_TAG is used in upstream:
* it is set for prelease and rcd builds only
* it overrides the result of `make tag`
* triggers strict integration tests, ignoring one test
* influences quay.io/rhacs-eng/stackrox-operator-bundle build tag
* prepend go-build directives to tests
* determines in e2e tests which images to pull
* used in dockerized builds make targets (which we don't use in GHA or Konflux)
* Setting the release GOTAGS if CI flag is also present, as we already know.

*Downstream:*

Set as env var to the CI_VERSION value in the builder stages.
* We don't always use make targets there, so even though we set CI=1 and BUILD_TAG=CI_VERSION, that doesn't result in `-tags release` automatically, which is why we need to specify GOTAGS=release explicitly (e.g. https://gitlab.cee.redhat.com/stackrox/rhacs-midstream/-/blob/rhacs-1.0-rhel-8/distgit/containers/rhacs-roxctl/Dockerfile.in?ref_type=heads#L7) in some cases.
* In other cases, like https://gitlab.cee.redhat.com/stackrox/rhacs-midstream/-/blob/rhacs-1.0-rhel-8/distgit/containers/rhacs-scanner-v4/Dockerfile.in?ref_type=heads#L15, we use make targets, and set CI+BUILD_TAG, which would set the release flag by itself already + override the TAG (not sure if it is used in this particular make target).

Konflux related:
* controls which scanner v4 vuln mapping thing is expected (to be set in ROX-23335)

> why we set [BUILD_TAG] in downstream builds

For Scanner v4, the BUILD_TAG is used for "logging and info" (https://redhat-internal.slack.com/archives/C05TS9N0S7L/p1731928556003139?thread_ts=1731671992.584229&cid=C05TS9N0S7L). 
Still no idea in most other cases, since we don't use make targets there (effect of "CI+BUILD_TAG=-tags release" doesn't happen) and always explicitly set GOTAGS=release. 

I cannot figure this part out on my own.

How to resolve this?

> We need Konflux builds to have release config because we'll be releasing them. Also, we don't need no-release, development builds from Konflux at this point.

> Therefore, we can take a shortcut and just make all Konflux builds everywhere to be with -tags release. Save time now and revise it later - that's [ROX-20240](https://issues.redhat.com/browse/ROX-20240).

We're on the same page about these points. I would suggest that we therefore go ahead with this PR and you help us answer the question of BUILD_TAG offline.

Reply:
The simple explanation for `BUILD_TAG` is if you do the following in the stackrox repo:
```bash
$ git checkout 4.5.5

$ make tag
4.5.5-0-g74f9ef246c
```

Think about this for a moment. You're on a _tagged_ commit, you run `make tag` and you get suffix (`-0-g74f9ef246c`).

Intentionally (I think yes) or unintentionally, it was made that local builds even for tagged commits have git-describe suffix so that you won't mess up release images even if you accidentally push the locally-built ones to container repos because true release images have image tags `4.5.5` and your local ones have `4.5.5-0-g74f9ef246c`.

And here's how it was meant that we are building _a release_ <https://github.com/stackrox/stackrox/blob/4a850371bcc0d78ba5ee6f0c1457ac19f8ee74b5/make/env.mk#L43-L48>
I.e. only in CI and only when our friend `BUILD_TAG` is set. Further, as you noticed `make tag` will return `BUILD_TAG` instead of doing `git describe` if `BUILD_TAG` is set. Set `BUILD_TAG` to 4.5.5 and `make tag` will return that.

You already saw examples how `BUILD_TAG` is set upstream for release and RC builds and how downstream sets `BUILD_TAG` effectively hardcoding it for each release. Basically, that was the idea that **CI** recognizes when it's a tagged commit and CI would inject `BUILD_TAG` variable in order to ensure `make tag` would return `4.5.5` without suffix.

In fact, `BUILD_TAG` was formerly called `RELEASE_TAG` but Gavin at some point did a big overhaul and I saw that in the mess of environment variables `RELEASE_TAG` makes things more confusing and suggested the rename. I suppose you felt how messy things are when you did your investigation around `BUILD_TAG`. I wish we stop introducing more environment variables as a vessel of passing things in CI.

----

We know that `BUILD_TAG` is used for `make tag` to not print the suffix but why is _that_ necessary?

This starts from https://github.com/stackrox/stackrox/blob/4a850371bcc0d78ba5ee6f0c1457ac19f8ee74b5/status.sh#L1-L7

I suggest you to trace where `status.sh` gets called and where `STABLE_MAIN_VERSION` gets injected to become aware of these workings.
Ultimately, this is about embedded versions which our Go code uses as image tags when generating installation manifests.
Embedded versions must match the image tags that were actually built and pushed (otherwise the thing won't start), therefore when we tag images as `4.5.5` we need to make sure that `STABLE_MAIN_VERSION` is also `4.5.5` and doesn't have the suffix.

There are three places that need these embedded versions:
- Central which uses embedded versions to generate Secured Cluster manifests in manifest-based installation.
- `roxctl` which uses embedded versions to generate Helm charts and to generate Central installation manifests.
- Operator which uses to generate Helm charts from which operands are deployed. Both in Konflux and downstream `RELATED_IMAGE_*` environment variables get populated and put in operator bundle CSV file. They effectively work as override to image tags and so it actually does not matter what are embedded versions for operators built in that context. However, the upstream-built operator doesn't have overrides and relies on these embedded versions.

----

> Embedded versions must match the actual image tags

is the primary reason why `BUILD_TAG` is populated (and it gets populated with the name of the git tag).
As you saw, it is populated in upstream and downstream builds, and populating it consistently in Konflux is part of my current task ROX-26026.

Looking again at these lines
https://github.com/stackrox/stackrox/blob/4a850371bcc0d78ba5ee6f0c1457ac19f8ee74b5/make/env.mk#L43-L48
I think the meaning was: if it's a tagged build in CI, that's a release build. And this meaning is valid still, although we have more things that define the release, or what _kind of_ release that is, e.g. branding.

In the Fast Stream future, I don't think we'll necessarily be tagging each commit that we'll want to release and deploy, and so the formula "if it's a tagged build in CI, that's a release build" won't apply. We'll need to force release build irrespective of git tags.

----

That was a lot of words and I did not tell what's my suggestion about `BUILD_TAG` in your current PR about `GOTAGS`, but maybe you made your conclusion from this?

Reply:
Also, if you have a feeling the way things are set up and chained is overcomplicated, I fully agree. As soon as we get rid of downstream releases, we'll have flexibility to change things. Just need to make sure we carefully write down use cases the current monstrosity supports and make sure we don't lose the ones that are important.

Reply:
> injecting BUILD_TAG it's done on the level above Makefiles and Dockerfiles. Therefore you'll find it being set above Makefiles and Dockerfiles. This was a design decision which predates me joining this project.

I think I was missing the location where this is done. AFAICT now, that is done in:
https://github.com/stackrox/stackrox/blob/master/scripts/ci/lib.sh#L1310-L1312

> In the Fast Stream future, I don't think we'll necessarily be tagging each commit that we'll want to release and deploy, and so the formula "if it's a tagged build in CI, that's a release build" won't apply. We'll need to force release build irrespective of git tags.

Yes, we need to solve this _later_.

> That was a lot of words and I did not tell what's my suggestion about BUILD_TAG in your current PR about GOTAGS, but maybe you made your conclusion from this?

* The whole thing is a mess and will need a refactoring once we're off CPaaS
* We need BUILD_TAG to be set and your PR https://github.com/stackrox/stackrox/pull/13422/ will eventually take care of it
* We can explicitly set GOTAGS=release. It will make builds on Konflux release type, but it may be redundant as we have the CI + BUILD_TAG combination.